Firestore/core/src/firebase/firestore/core/query.h

```diff
@@ -36,28 +42,46 @@ class Query {
    * @return A new instance of Query.
    */
   static Query AtPath(model::ResourcePath path) {
-    return Query(std::move(path));
+    return Query(std::move(path), std::vector<std::unique_ptr<core::Filter>>());
   }
 
   /** Initializes a query with all of its components directly. */
-  explicit Query(model::ResourcePath path /* TODO(rsgowman): other params */)
-      : path_(std::move(path)) {
-  }
+  Query(model::ResourcePath path,
+        std::vector<std::unique_ptr<core::Filter>>&&
+            filters /* TODO(rsgowman): other params */);
 
+  /** The base path of the query. */
   const model::ResourcePath& path() const {
     return path_;
   }
 
   /** Returns true if the document matches the constraints of this query. */
   bool Matches(const model::Document& doc) const;
 
+  /**
+   * Returns a copy of this Query object with the additional specified filter.
+   */
+  Query Filter(std::unique_ptr<core::Filter> filter) const;
+
  private:
+  Query(model::ResourcePath path,
+        std::vector<std::shared_ptr<core::Filter>>
+            filters /* TODO(rsgowman): other params */)
+      : path_(std::move(path)), filters_(std::move(filters)) {
+  }
+
   bool MatchesPath(const model::Document& doc) const;
   bool MatchesFilters(const model::Document& doc) const;
   bool MatchesOrderBy(const model::Document& doc) const;
   bool MatchesBounds(const model::Document& doc) const;
 
   const model::ResourcePath path_;
+
+  // Filters are shared across related Query instance. i.e. when you call
+  // Query::Filter(f), a new Query instance is created that contains all of the
+  // existing filters, plus the new one. (Both Query and Filter objects are
+  // immutable.) Filters are not shared across unrelated Query instances.
+  const std::vector<std::shared_ptr<core::Filter>> filters_;
```

Comment:
Just to throw it out, an alternative would be to hide the shared ownership inside `Filter` -- have it store a `shared_ptr` to implementation, which would allow storing plain `Filter`s in the vector (like e.g. [DocumentKey](https://github.com/firebase/firebase-ios-sdk/blob/04a28fce81c737b6505e6c542a14d8529c9f891d/Firestore/core/src/firebase/firestore/model/document_key.h#L98-L100) does). Feel free to push back on this.

Reply:
It's unclear to me if this is worth the extra layer... but it *does* simplify the Query object. Here's a quick implementation (which may or may not actually be correct! but I imagine it'd look something like this): https://github.com/firebase/firebase-ios-sdk/pull/1497

WDYT?

Reply:
Let's continue the discussion in the new PR (thanks for doing this!).